ci/pipeline/ray_coverage.py

```diff
@@ -0,0 +1,110 @@
+import logging
+import os
+import subprocess
+import sys
+from datetime import date
+
+import click
+import boto3
+
+
+def _get_logger():
+    logging.basicConfig(
+        stream=sys.stderr,
+        level=logging.INFO,
+        format="%(asctime)s:%(levelname)s:%(name)s:%(message)s",
+    )
+    logger = logging.getLogger()
+    logger.setLevel(logging.INFO)
+    return logger
+
+
+_logger = _get_logger()
+_COVERAGE_FILE_NAME = "ray_release.cov"
+_S3_BUCKET_NAME = "ray-test-coverage"
+_S3_BUCKET_DIR = "ci"
+
+
+@click.command()
+@click.argument("test_target", required=True, type=str)
+@click.option(
+    "--artifact-dir",
+    default="/artifact-mount",
+    type=str,
+    help=(
+        "This directory is used to store artifacts, such as coverage information. "
+        "In buildkite CI, this is usually artifact-mount."
+    ),
+)
+@click.option(
+    "--upload",
+    is_flag=True,
+    show_default=True,
+    default=False,
+    help=("Upload the computed coverage data to S3."),
+)
+def main(
+    test_target: str,
+    artifact_dir: str = "/artifact-mount",
+    upload: bool = False,
+) -> None:
+    """
+    This script collects dynamic coverage data for the test target, and upload the
+    results to database (S3).
+    """
+    _logger.info(f"Collecting coverage for test target: {test_target}")
+    coverage_file = os.path.join(artifact_dir, _COVERAGE_FILE_NAME)
+    _run_test(test_target, coverage_file)
+    coverage_info = _collect_coverage(coverage_file)
+    _logger.info(coverage_info)
+    if upload:
+        _upload_coverage_info(coverage_file)
+    return 0
+
+
+def _upload_coverage_info(coverage_file: str) -> None:
+    s3_file_name = (
+        f"{_S3_BUCKET_DIR}/ray-release-{date.today().strftime('%Y-%m-%d')}.cov"
```

Comment:
maybe add a random nounce or something? so that run from the same day do not overwrite each other?

maybe also consider create an associated meta file with git commit digest and buildkite build id?